### PR TITLE
Hans data

### PR DESCRIPTION
**Pull request by sgugger**
This is the first step toward solving #4742: to be able to use the Trainer API, we first need to remove the TensorDataset and have datasets with dict items. This PR addresses that and updates the training and evaluation script accordingly.

It takes the multiple choice as a reference implementation, using the same file structure (hence the removal of "hans_processor.py") and implements in "utils_hans.py":
- a `HansDataset` and a `TFHansDataset` that implement the logic of the old method `load_and_cache_examples`
- a `HansProcessor` (copied from before)
- a `hans_convert_examples_to_features` with the same logic as before but using the tokenizer method for padding instead of re-implementing it.

Side question: it doesn't look like the `TFMultipleChoiceDataset` I use as a reference for this implementation uses caching, maybe it should be added in the future?

### CODE REVIEW

File: examples/adversarial/test_hans.py [hunk 1]
```diff
@@ -154,10 +158,14 @@ def train(args, train_dataset, model, tokenizer):
...
+            inputs = {
+                "input_ids": batch["input_ids"],
+                "attention_mask": batch["attention_mask"],
+                "labels": batch["labels"],
+            }
             if args.model_type != "distilbert":
                 inputs["token_type_ids"] = (
-                    batch[2] if args.model_type in ["bert", "xlnet"] else None
+                    batch["token_type_ids"] if args.model_type in ["bert", "xlnet"] else None
                 )  # XLM, DistilBERT and RoBERTa don't use segment_ids
```
Comment (LysandreJik) [inline examples/adversarial/test_hans.py:169]: This can now be way simpler. Models that don't use `token_type_ids` don't output it anymore (I believe XLM makes use of them, but I may be wrong).

Therefore by using `tokenizer.encode_plus`, you can obtain the model-specific output right away.

e.g. with BERT:

```py
bert_tokenizer.encode_plus("hey, how are you?")
# {'input_ids': [101, 23998, 117, 1293, 1132, 1128, 136, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1]}
```

whereas with DistilBERT:

```py
distilbert_tokenizer.encode_plus("hey, how are you?")
# {'input_ids': [101, 23998, 117, 1293, 1132, 1128, 136, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1]}
```

This is so that you can chain the two methods:

```py
inputs = tokenizer.encode_plus(xxx)
model(**inputs)
```

File: examples/adversarial/test_hans.py [hunk 2]
```diff
@@ -264,10 +284,14 @@ def evaluate(args, model, tokenizer, prefix=""):
...
-                inputs = {"input_ids": batch[0], "attention_mask": batch[1], "labels": batch[3]}
+                inputs = {
+                    "input_ids": batch["input_ids"],
+                    "attention_mask": batch["attention_mask"],
+                    "labels": batch["labels"],
+                }
                 if args.model_type != "distilbert":
                     inputs["token_type_ids"] = (
-                        batch[2] if args.model_type in ["bert", "xlnet"] else None
+                        batch["token_type_ids"] if args.model_type in ["bert", "xlnet"] else None
```
Comment (LysandreJik) [inline examples/adversarial/test_hans.py:294]: Same here

**Review by LysandreJik — COMMENTED** [review]
Great, thanks for looking into it! There was a legacy piece of code that could be improved, I left you a comment.

Regarding caching, how long does it take to tokenize the dataset with fast tokenizers? If it's very fast it may not be necessary. It wouldn't do any harm to implement it anyway, though!

**Review by LysandreJik — APPROVED** [review]
Looks good!

File: examples/adversarial/utils_hans.py [hunk 3]
```diff
@@ -14,108 +14,339 @@
...
+            max_length=max_length,
+            pad_to_max_length=True,
+            return_overflowing_tokens=True,
+        )
+        if "num_truncated_tokens" in inputs and inputs["num_truncated_tokens"] > 0:
+            logger.info(
+                "Attention! you are cropping tokens (swag task is ok). "
+                "If you are training ARC and RACE and you are poping question + options,"
+                "you need to try to use a bigger max seq length!"
+            )
```
Comment (julien-c) [inline examples/adversarial/utils_hans.py:321]: I think this warning is non-relevant here

File: examples/adversarial/utils_hans.py [hunk 4]
```diff
@@ -14,108 +14,339 @@
...
+
+        if output_mode == "classification":
+            label = label_map[example.label] if example.label in label_map else 0
+        elif output_mode == "regression":
+            label = float(example.label)
+        else:
+            raise KeyError(output_mode)
```
Comment (julien-c) [inline examples/adversarial/utils_hans.py:328]: Only classification here so I think you can remove this (and the `output_mode` arg)

File: examples/adversarial/test_hans.py [hunk 5]
```diff
@@ -25,13 +25,10 @@
 
 import numpy as np
 import torch
-from torch.utils.data import DataLoader, RandomSampler, SequentialSampler, TensorDataset
```
Comment (julien-c) [inline examples/adversarial/test_hans.py:28]: not related to this PR, but in the next maybe we can rename this file to `run_hans.py` like other examples

File: examples/adversarial/utils_hans.py [hunk 6]
```diff
@@ -14,108 +14,339 @@
...
+                    logger.info(f"Creating features from dataset file at {data_dir}")
+                    label_list = processor.get_labels()
+
+                    if task in ["mnli", "mnli-mm"] and tokenizer.__class__ in (
+                        RobertaTokenizer,
+                        RobertaTokenizerFast,
+                        XLMRobertaTokenizer,
+                    ):
+                        # HACK(label indices are swapped in RoBERTa pretrained model)
+                        label_list[1], label_list[2] = label_list[2], label_list[1]
```
Comment (julien-c) [inline examples/adversarial/utils_hans.py:136]: All this can be removed too AFAICT

File: examples/adversarial/utils_hans.py [hunk 7]
```diff
@@ -14,108 +14,339 @@
...
+    """Processor for the HANS data set."""
+
+    def get_example_from_tensor_dict(self, tensor_dict):
+        """See base class."""
+        return InputExample(
+            tensor_dict["idx"].numpy(),
+            tensor_dict["premise"].numpy().decode("utf-8"),
+            tensor_dict["hypothesis"].numpy().decode("utf-8"),
+            str(tensor_dict["label"].numpy()),
+        )
```
Comment (julien-c) [inline examples/adversarial/utils_hans.py:250]: This is not used anymore AFAICT – you can remove

**Review by julien-c — COMMENTED** [review]
Added a few comments, otherwise LGTM
